imap_processing/ultra/l1c/ultra_l1c_pset_bins.py

```diff
@@ -44,11 +48,101 @@ def build_spatial_bins(spacing: float = 0.5) -> tuple[np.ndarray, np.ndarray]:
         Array of azimuth bin boundary values.
     el_bin_edges : np.ndarray
         Array of elevation bin boundary values.
+    az_bin_midpoints : np.ndarray
+        Array of azimuth bin midpoint values.
+    el_bin_midpoints : np.ndarray
+        Array of elevation bin midpoint values.
     """
     # Azimuth bins from 0 to 360 degrees.
     az_bin_edges = np.arange(0, 360 + spacing, spacing)
+    az_bin_midpoints = az_bin_edges[:-1] + spacing / 2  # Midpoints between edges
 
     # Elevation bins from -90 to 90 degrees.
     el_bin_edges = np.arange(-90, 90 + spacing, spacing)
+    el_bin_midpoints = el_bin_edges[:-1] + spacing / 2  # Midpoints between edges
+
+    return az_bin_edges, el_bin_edges, az_bin_midpoints, el_bin_midpoints
+
+
+def cartesian_to_spherical(
+    v: tuple[np.ndarray, np.ndarray, np.ndarray],
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Convert cartesian coordinates to spherical coordinates.
+
+    Parameters
+    ----------
+    v : tuple[np.ndarray, np.ndarray, np.ndarray]
+        The x,y,z-components of the velocity vector.
+
+    Returns
+    -------
+    az : np.ndarray
+        The azimuth angles in degrees.
+    el : np.ndarray
+        The elevation angles in degrees.
+    r : np.ndarray
+        The radii, or magnitudes, of the vectors.
+    """
+    vx, vy, vz = np.hsplit(v, 3)
+    vx, vy, vz = vx.flatten(), vy.flatten(), vz.flatten()  # Flatten the arrays
+
+    # Magnitude of the velocity vector
+    magnitude_v = np.sqrt(vx**2 + vy**2 + vz**2)
+
+    vhat_x = -vx / magnitude_v
+    vhat_y = -vy / magnitude_v
+    vhat_z = -vz / magnitude_v
+
+    # Convert from cartesian to spherical coordinates (azimuth, elevation)
+    # Radius (magnitude)
+    r = np.sqrt(vhat_x**2 + vhat_y**2 + vhat_z**2)
```

Comment:
Since `v_hat` is normalized, `r` should/will always be 1. The correct radius should be `magnitude_v`, no?

Reply:
NIce catch.